File: pistoleiros/salao/salao2.py
import sys
import time


mapSize = int(sys.argv[1])
b = int(sys.argv[2])
c = int(sys.argv[3])


tabuleiro = [["_" for _ in range(mapSize)] for _ in range(mapSize)]
tabuleiros = set()


def salao(b, c, map, solutions):
    if b == 0 and c == 0:
        if checaTabuleiro(map):
            solutions.add(tuple(tuple(row) for row in map))
            """  for row in map:
                print(" ".join(row))
            print("\n") """
        return
    
    if b < 0 or c < 0:
        return

    #adding B
    for i in range(len(map)):
        for j in range(len(map)):
            if map[i][j] =="_" and notContiguous(i, j, map):
                map[i][j] = "b"
                salao(b-1, c, map, solutions)
                map[i][j] = "_"

    for i in range(len(map)):
        for j in range(len(map)):
            if map[i][j]=="_" and isValid(i, j, map, "c", "b"):
              
                map[i][j] = "c"
                salao(b, c-1, map, solutions)
           
                map[i][j] = "_"


    return solutions


def notContiguous(l, c, map):
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1),
                  (-1, -1), (-1, 1), (1, -1), (1, 1)]
    for dr, dc in directions:
        if 0 <= l + dr < len(map) and 0 <= c + dc < len(map):
            if map[l + dr][c + dc] == "b":
                return False
    return True


def isValid(l, c, map, pistoleiro, enemy):
    enemySeen = 0
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1),
                  (-1, -1), (-1, 1), (1, -1), (1, 1)]
    for dr, dc in directions:
        for step in range(1, mapSize):
            r, d = l + dr * step, c + dc * step
            if not (0 <= r < mapSize and 0 <= d < mapSize):
                break
            if map[r][d] == pistoleiro:
                return False
            if map[r][d] == enemy:
                enemySeen += 1
                break


    return enemySeen >= 2


def checaTabuleiro(map):
    for i in range(len(map)):
        for j in range(len(map)):
            if map[i][j] == "b":
                if not isValid(i, j, map, "b", "c"):
                    return False
    return True


inicio = time.time() 
soluçoes = salao(b, c, tabuleiro, tabuleiros)
print(len(soluçoes))
fim = time.time() 

def salvar_solucoes_em_txt(solutions, nome_arquivo="solucoes2.txt"):
    with open(nome_arquivo, "w") as f:
        for solution in solutions:
            for row in solution:
                f.write(" ".join(row) + "\n")
            f.write("\n")  

salvar_solucoes_em_txt(soluçoes)
tempo_execucao = fim - inicio
print(f"Tempo de execução: {tempo_execucao:.6f} segundos")

File: pistoleiros/salao/salao2.java
import java.util.HashSet;
import java.util.Set;

public class salao2 {

    public static void main(String[] args) {

        long tempoInicial = System.currentTimeMillis();

        int mapSize = Integer.parseInt(args[0]);
        int b = Integer.parseInt(args[1]);
        int c = Integer.parseInt(args[2]);

        char[][] board = new char[mapSize][mapSize];
        for (int i = 0; i < mapSize; i++) {
            for (int j = 0; j < mapSize; j++) {
                board[i][j] = '_';
            }
        }

        Set<String> solutions = new HashSet<>();
        System.out.println(salao(b, c, board, solutions, true)); // Começa com 'b'
        long tempoFinal = System.currentTimeMillis();
        long tempoTotal = tempoFinal - tempoInicial;
        System.out.println("Tempo total de execução: " + tempoTotal + " milissegundos");

    }

    private static int salao(int b, int c, char[][] map, Set<String> solutions, boolean placeB) {
        if (b == 0 && c == 0) {
            if (checaTabuleiro(map)) {
                solutions.add(boardToString(map));
            }
            return solutions.size();
        }

        for (int i = 0; i < map.length; i++) {
            for (int j = 0; j < map.length; j++) {
                if (map[i][j] == '_') {
                    if (placeB && b > 0 && isValid(i, j, map, 'b', 'c')) {
                        map[i][j] = 'b';
                        salao(b - 1, c, map, solutions, false); // Alterna para 'c'
                        map[i][j] = '_';
                    } else if (!placeB && c > 0 && isValid(i, j, map, 'c', 'b')) {
                        map[i][j] = 'c';
                        salao(b, c - 1, map, solutions, true); // Alterna para 'b'
                        map[i][j] = '_';
                    }
                }
            }
        }

        return solutions.size();
    }

    private static boolean isValid(int l, int c, char[][] map, char pistoleiro, char enemy) {
        int[][] directions = {
                { -1, 0 }, { 1, 0 }, { 0, -1 }, { 0, 1 },
                { -1, -1 }, { -1, 1 }, { 1, -1 }, { 1, 1 }
        };

        for (int[] dir : directions) {
            for (int step = 1; step < map.length; step++) {
                int newRow = l + dir[0] * step;
                int newCol = c + dir[1] * step;

                if (newRow < 0 || newRow >= map.length || newCol < 0 || newCol >= map.length) {
                    break;
                }
                if (map[newRow][newCol] == pistoleiro) {
                    return false;
                }
                if (map[newRow][newCol] == enemy) {
                    break;
                }
            }
        }
        return true;
    }

    private static boolean checaTabuleiro(char[][] map) {
        for (int i = 0; i < map.length; i++) {
            for (int j = 0; j < map.length; j++) {
                if (map[i][j] == 'b' || map[i][j] == 'c') {
                    if (!contaInimigos(i, j, map, map[i][j])) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    private static boolean contaInimigos(int l, int c, char[][] map, char pistoleiro) {
        char enemy = pistoleiro == 'b' ? 'c' : 'b';
        int enemySeen = 0;
        int[][] directions = {
                { -1, 0 }, { 1, 0 }, { 0, -1 }, { 0, 1 },
                { -1, -1 }, { -1, 1 }, { 1, -1 }, { 1, 1 }
        };

        for (int[] dir : directions) {
            for (int step = 1; step < map.length; step++) {
                int newRow = l + dir[0] * step;
                int newCol = c + dir[1] * step;

                if (newRow < 0 || newRow >= map.length || newCol < 0 || newCol >= map.length) {
                    break;
                }
                if (map[newRow][newCol] == enemy) {
                    enemySeen++;
                    break;
                }
                if (map[newRow][newCol] == pistoleiro) {
                    break;
                }
            }
        }

        return enemySeen >= 2;
    }

    private static String boardToString(char[][] map) {
        StringBuilder sb = new StringBuilder();
        for (char[] row : map) {
            for (char cell : row) {
                sb.append(cell);
            }
            sb.append('\n');
        }
        return sb.toString();
    }
}
